src/service/resource.js

```diff
@@ -1032,26 +1032,6 @@ export class Resource {
     return this.loadedOnce_;
   }
 
-  /**
-   * Whether the resource is currently visible in the viewport.
-   * @return {boolean}
-   */
-  isInViewport() {
-    const isInViewport = this.element.isInViewport();
-    if (isInViewport) {
-      this.resolveDeferredsWhenWithinViewports_();
```

Comment:
Did we already remove `whenWithinViewport`?

Reply:
This is something I haven't figured out yet.
I don't think we necessarily need to remove `whenWithinViewport` since it doesn't rely on inViewport at all.

For some reason `isInViewport()` ends up having a side effect which checks whether all the current in-flight promises deserve to be resolved. We may need to trigger this somehow else

Reply:
Hmm. Per reading this code it looks like `whenWithinViewport` is computed via `inViewport`. And, yes, you're right - it's implemented really poorly. IMHO, we can just support `whenWithinViewport` either via the same `observe` we already have (by calling `unobserve` as soon as `observe` has happened) or via a separate `IntersectionObserver`. It depends on the uses, how many of them, etc.

Reply:
Unfortunately `whenWithinViewport` takes in a parameter that dictates how many viewports away it is.
The helper is only used in two places in ads. One is a race condition fix, and the other is to delay ad loading.
I'll separate out the research into its own issue/pr.